chain/ethereum/ethereum.go

```diff
@@ -173,6 +174,54 @@ func ChunkedFilterLogs(ctx context.Context, client ETHClient, addresses []common
 	return failedQueries, err
 }
 
+// BatchedFilterLogs queries the blockchain for past events in batches.
+// It slices addresses and total number of blocks with pre-defined batch size.
+// The results are later fed into a log chan that was provided by the caller.
+// Failed query intervals are retried after a backoff period, increased exponentially.
+func BatchedFilterLogs(ctx context.Context, client ETHClient, addresses []common.Address, fromBlock, toBlock uint64, logChan chan<- types.Log, backoff uint) {
+	// Split filterlog queries into 7350 contracts due to json-rpc limitation (undocumented)
+	const (
+		addressBatchSize int    = 7350
+		blockBatchSize   uint64 = 50
+	)
+
+	if backoff == 0 {
+		backoff = 1
+	}
+
+	if len(addresses) > addressBatchSize {
+		BatchedFilterLogs(ctx, client, addresses[:addressBatchSize], fromBlock, toBlock, logChan, backoff)
+		BatchedFilterLogs(ctx, client, addresses[addressBatchSize:], fromBlock, toBlock, logChan, backoff)
+		return
+	}
+
+	if toBlock-fromBlock > blockBatchSize {
+		BatchedFilterLogs(ctx, client, addresses, toBlock-blockBatchSize, toBlock, logChan, backoff)
+		BatchedFilterLogs(ctx, client, addresses, fromBlock, toBlock-blockBatchSize-1, logChan, backoff)
+		return
+	}
+
+	log.Debug().Uint64("from", fromBlock).Uint64("to", toBlock).Msg("retrieving historical events...")
+
+	query := ethereum.FilterQuery{
+		FromBlock: big.NewInt(int64(fromBlock)),
+		ToBlock:   big.NewInt(int64(toBlock)),
+		Addresses: addresses,
+	}
+
+	logs, err := client.FilterLogs(ctx, query)
+	if err != nil {
+		log.Warn().Err(err).Uint64("from", fromBlock).Uint64("to", toBlock).Uint("backoff minutes", backoff).Msg("retrying interval")
+		time.Sleep(time.Duration(backoff) * time.Minute)
```

Comment:
It should be like this or the function will run forever without a way to exit:
```
	select {
	case <-time.After(time.Duration(backoff) * time.Minute):
		BatchedFilterLogs(ctx, client, addresses, fromBlock, toBlock, logChan, backoff<<1)
		return
	case <-ctx.Done():
		// Context timeout => Exit
	}
```

Reply:
One minute is quite a long duration. I suggest to reduce it to 1 second instead.

Reply:
Since the error is caused by RPC provider, I think it is better to give provider longer time to heal and make fewer calls. Also this is an ongoing operation in the background thread so it won't effect the connector.